File: src/main/java/dev/shota/decompiler/window/menu/file/FileCloseTab.java
package dev.shota.decompiler.window.menu.file;

import dev.shota.decompiler.window.utils.Language;
import dev.shota.decompiler.window.viewer.Viewer;
import javafx.application.Platform;
import javafx.collections.ListChangeListener;
import javafx.scene.control.Tab;
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;

public class FileCloseTab extends JMenuItem implements ActionListener {

    public FileCloseTab() {
        setText(Language.get("file.closeTab"));
        setMnemonic(KeyEvent.VK_W);
        setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_W, Toolkit.getDefaultToolkit().getMenuShortcutKeyMaskEx()));
        addActionListener(this);
        setEnabled(false);

        Platform.runLater(() -> Viewer.INSTANCE.getTabs().addListener((ListChangeListener<Tab>) c -> setEnabled(!Viewer.INSTANCE.getSelectionModel().isEmpty())));
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        Platform.runLater(() -> Viewer.INSTANCE.getTabs().remove(Viewer.INSTANCE.getSelectionModel().getSelectedItem()));
    }

}

File: jvm/src/main/kotlin/moe/sota/decompiler/jvm/menus/file/FileCloseTab.kt
package moe.sota.decompiler.jvm.menus.file

import com.formdev.flatlaf.extras.components.FlatMenuItem
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.KeyEvent
import javax.swing.KeyStroke
import moe.sota.decompiler.jvm.controllers.TabsController
import moe.sota.decompiler.jvm.services.LanguageService

class FileCloseTab(languageService: LanguageService, private val tabsController: TabsController) :
    FlatMenuItem(), ActionListener {

    init {
        accelerator =
            KeyStroke.getKeyStroke(KeyEvent.VK_W, Toolkit.getDefaultToolkit().menuShortcutKeyMaskEx)
        isEnabled = false
        mnemonic = KeyEvent.VK_W
        text = languageService.getString("file.closeTab")

        addActionListener(this)
    }

    override fun actionPerformed(e: ActionEvent?) {
        tabsController.closeTab()
    }
}
